tests/components/sftp_storage/test_init.py

```diff
@@ -0,0 +1,116 @@
+"""Tests for SFTP Storage."""
+
+from pathlib import Path
+from unittest.mock import patch
+
+from asyncssh.sftp import SFTPPermissionDenied
+import pytest
+
+from homeassistant.components.sftp_storage import SFTPConfigEntryData
+from homeassistant.components.sftp_storage.const import DOMAIN
+from homeassistant.config_entries import ConfigEntryState
+from homeassistant.core import HomeAssistant
+
+from .asyncssh_mock import SSHClientConnectionMock
+from .conftest import (
+    CONF_HOST,
+    CONF_PRIVATE_KEY_FILE,
+    CONF_USERNAME,
+    USER_INPUT,
+    ComponentSetup,
+)
+
+
+@pytest.mark.usefixtures("mock_ssh_connection")
+async def test_setup_and_unload(
+    hass: HomeAssistant,
+    setup_integration: ComponentSetup,
+    caplog: pytest.LogCaptureFixture,
+) -> None:
+    """Test successful setup and unload."""
+
+    # Patch the `exists` function of Path so that we can also
+    # test the `homeassistant.components.sftp_storage.client.get_client_keys()` function
+    with (
+        patch(
+            "homeassistant.components.sftp_storage.client.SSHClientConnectionOptions"
+        ),
+        patch("pathlib.Path.exists", return_value=True),
+    ):
+        await setup_integration()
+
+    entries = hass.config_entries.async_entries(DOMAIN)
+    assert len(entries) == 1
+    assert entries[0].state is ConfigEntryState.LOADED
+
+    await hass.config_entries.async_unload(entries[0].entry_id)
+
+    assert entries[0].state is ConfigEntryState.NOT_LOADED
+    assert (
+        f"Unloading {DOMAIN} integration for host {entries[0].data[CONF_USERNAME]}@{entries[0].data[CONF_HOST]}"
+        in caplog.messages
+    )
+
+
+async def test_setup_error(
+    mock_ssh_connection: SSHClientConnectionMock,
+    hass: HomeAssistant,
+    setup_integration: ComponentSetup,
+) -> None:
+    """Test setup error."""
+    mock_ssh_connection._sftp._mock_chdir.side_effect = SFTPPermissionDenied(
+        "Error message"
+    )
+    await setup_integration()
+
+    entries = hass.config_entries.async_entries(DOMAIN)
+    assert len(entries) == 1
+    assert entries[0].state is ConfigEntryState.SETUP_ERROR
+
+
+async def test_setup_unexpected_error(
+    hass: HomeAssistant,
+    setup_integration: ComponentSetup,
+    caplog: pytest.LogCaptureFixture,
+) -> None:
+    """Test setup error."""
+    with patch(
+        "homeassistant.components.sftp_storage.client.connect",
+        side_effect=OSError("Error message"),
+    ):
+        await setup_integration()
+
+    entries = hass.config_entries.async_entries(DOMAIN)
+    assert len(entries) == 1
+    assert entries[0].state is ConfigEntryState.SETUP_ERROR
+    assert (
+        "Failure while attempting to establish SSH connection. Please check SSH credentials and if changed, re-install the integration"
+        in caplog.text
+    )
+
+
+async def test_async_remove_entry(
```

Comment:
Since we're being destructive when removing config entries, let's let this test start with two config entries. After the 1st entry is removed, that entries private key should not be removed. After the second config entry is removed, both key files and the containing folder should be removed.